### ISSUE TITLE
Allow object arg to `Session.set`

### ISSUE THREAD

**Issue by lorensr** [meteor/meteor#3574]
[FR] Sometimes I have long lists of:

```js
Session.set 'foo', yes
Session.set 'bar', 0
Session.set 'foobar', 'hi'
```

Would be nice to overload first arg so I could set multiple at once:

```js
Session.set 
  foo: yes
  bar: 0
  foobar: 'hi'
```

**Comment by awatson1978** [comment]
There is a [session-extended-api](https://github.com/awatson1978/session-extended-api) package that provides ``toggle()`` and ``clear()`` functions, and is intended as a place to try these kinds of requests.  We could take a look at adding this functionality there.  I had been planning on updating it this week and making sure it was , and could implement this fairly quickly.  

The biggest issue with this proposal that I see is that ``Session.set()`` already takes an object (EJSON, I think), and disambiguating between setting an object to a variable and setting multiple variables.

I'd be prone to follow the ``Session.setDefault()`` naming schema, and implement this as ``Session.setMultiple()``, like so:

````js
Session.setMultiple({
  foo: true,
  bar: 0,
  foobar: 'hi'
});
````

And you should be able to get the above syntax via the coffeescript compiler.

**Comment by stubailo** [comment]
Thanks for the feature request! You might be interested in reading our [feature request guidelines](https://github.com/meteor/meteor/wiki/Contributing-to-Meteor#feature-requests).

**Comment by awatson1978** [comment]
@stubailo - now that Meteor 1.0 is out, and MDG is accepting feature requests again, will you all be taking pull requests on these kinds of things?  I'll probably be rolling into this tonight or tomorrow, since I need similar functionality for my own projects and it's a 10 minute fix and I'm in a place to just implement it.  But if you all will accept the pull-request against the ``session`` package, I'll move it from the ``session-extended-api`` and include the ``clear()`` and ``toggle()`` functions also.

**Comment by stubailo** [comment]
I would probably look at a PR for this, implemented on ReactiveDict (which is the base for Session). I think a good name to call it might be `Session.setAll` (ReactiveDict#setAll).

I've also found many use cases in my work for `clear` and `toggle`, so I would be in favor of having those as well.

I'm building a new API for local template state (instead of attaching individual reactive vars) which will also be based on `reactive-dict`, so it will be great to have all of these new shared features.

**Comment by awatson1978** [comment]
Implemented on ReactiveDict, huh?  Okay.  I think I could handle that.  Nothing too fancy.  Just keep it simple.  

What about testing?  I don't see any tinytests on the ``reactive-dict`` package, but they are plenty in the ``session`` package.

**Comment by lorensr** [comment]
I prefer overloading `set` versus making a new method. It appears programmers agree :)

http://programmers.stackexchange.com/questions/132369/should-we-rename-overloaded-methods

**Comment by stubailo** [comment]
For now, all of the tests are still in the session package. put your new tests there for now - I will move those soon since I'm doing some reactive-dict refactoring.

**Comment by stubailo** [comment]
Looks like Backbone overloads `set`, so maybe that will be a good idea. http://backbonejs.org/#Model-set

What are the chances people will accidentally pass the wrong thing? We should check for two things:

1. The first argument is an object
2. The second argument doesn't exist

**Comment by awatson1978** [comment]
Okidoki.  Got something pieced together, but there's a version constraint keeping the tests from passing.  Should I use publish-release and modify the default package versions to run the tests?  Or would it be better to maybe move the tinytests from ``session_tests.js`` to ``reactive-dict_tests.js`` so there's not the cross-package dependency on running the tests?

**Comment by stubailo** [comment]
Have you tried running the tests using a checkout of meteor instead of a release? AFAIK then there shouldn't be any version issues.

**Comment by awatson1978** [comment]
That's certainly what I thought I was doing.  Cloned meteor to my own account, downloaded, checkout a new branch, edited and tested with ``sudo meteor test-packages packages/session/``.  It finds the tests in the test-packages.  It just says that ``undefined is not a function`` in reference to the new ``.toggle()`` and ``.clear()`` methods and when it tries to parse the object of key/values.  Trying to specify version constraints in the ``package.js`` file told me that the constraints come from a <top level> meteor-platform@1.2.1.

**Comment by stubailo** [comment]
First of all, please don't ever use `sudo` to run `meteor`.

I think this workflow will work:

```
cd <meteor-checkout>
./meteor test-packages session
```

The important thing is that you are running the local `meteor` script from the checkout, not the global installed `meteor`. You do this by adding the `./` in front.

**Comment by awatson1978** [comment]
Aaaaaaah.  The ``./`` did the trick!  All the tests are passing green now!  :)

As for the not using ``sudo``, I'd love to be able to not use it, and I've been systematically removing it from all of my scripts and documents in the cookbook.  Unfortunately, in practice, meteor file permissions can get out of sync fairly easily when installing other utilities, libraries, etc.  (Right now, if I don't use ``sudo``, meteor burps errors about npm-shrinkwrap, css-parse, isopacket, etc.  Ug.)   Meteor really needs a file permissions audit script and a best practice on which directories need ``chown`` vs ``chmod`` permissions.    But that's a separate issue....

**Comment by stubailo** [comment]
yeah let's discuss this in a different place than this issue.

**Comment by glasser** [comment]
@stubailo should this issue be closed?

**Comment by stubailo** [comment]
Yes, it should!

**Comment by lorensr** [comment]
Landed in Meteor `v1.0.4`: https://github.com/meteor/meteor/blob/devel/History.md#tracker-1
